sources/imdsclient/src/lib.rs

```diff
@@ -16,26 +16,34 @@ The result is returned as a `String` _(ex. m5.large)_.
 
 #![deny(rust_2018_idioms)]
 
+use std::sync::{Arc, RwLock};
+
 use http::StatusCode;
 use log::{debug, info, trace, warn};
 use reqwest::Client;
 use serde_json::Value;
 use snafu::{ensure, OptionExt, ResultExt};
-use std::time::Duration;
-use tokio::time;
+use tokio::time::{timeout, Duration};
+use tokio_retry::{strategy::FibonacciBackoff, Retry};
 
 const BASE_URI: &str = "http://169.254.169.254";
 const PINNED_SCHEMA: &str = "2021-01-03";
 
 // Currently only able to get fetch session tokens from `latest`
 const SESSION_TARGET: &str = "latest/api/token";
 
+fn retry_strategy() -> impl Iterator<Item = Duration> {
+    // Retries (relative to start) at 0.5s, 1s, 2s, 3.5s, 6s, 10s, 16.5s, and then every 10s after.
+    FibonacciBackoff::from_millis(500).max_delay(Duration::from_secs(10))
+}
+
 /// A client for making IMDSv2 queries.
 /// It obtains a session token when it is first instantiated and is reused between helper functions.
 pub struct ImdsClient {
     client: Client,
     imds_base_uri: String,
-    session_token: String,
+    retry_timeout: Duration,
+    session_token: Arc<RwLock<String>>,
```

Comment:
`Arc` may not be needed since the ImdsClient is not `Clone`.

Reply:
I think you are correct. Everything seems to work with just RwLock<String>.